File: apps/flutter/lib/features/debts/domain/entities/debt.dart
class Debt {
  final String id;
  final String name;
  final String type;
  final String currency;
  final double currentBalance;
  final int? dueDay;
  final double? minimumPayment;
  final double? interestRateAnnual;
  final String? linkedAccountId;
  final String? paymentAccountId;
  final double? initialBalance;
  final bool isActive;
  final double amountDue;
  final double creditBalance;

  const Debt({
    required this.id,
    required this.name,
    required this.type,
    this.linkedAccountId,
    this.paymentAccountId,
    required this.currency,
    required this.currentBalance,
    required this.dueDay,
    required this.minimumPayment,
    required this.interestRateAnnual,
    this.initialBalance,
    required this.isActive,
    required this.amountDue,
    required this.creditBalance,
  });

  factory Debt.fromJson(Map<String, dynamic> json) {
    return Debt(
      id: (json["debtId"] ?? json["id"]) as String,
      name: json["name"] as String,
      type: json["type"] as String,
      linkedAccountId: json["linkedAccountId"] as String?,
      paymentAccountId: json["paymentAccountId"] as String?,
      currency: json["currency"] as String? ?? "BRL",
      currentBalance: (json["currentBalance"] as num?)?.toDouble() ?? 0,
      dueDay: json["dueDay"] as int?,
      minimumPayment: (json["minimumPayment"] as num?)?.toDouble(),
      interestRateAnnual: (json["interestRateAnnual"] as num?)?.toDouble(),
      initialBalance: (json["initialBalance"] as num?)?.toDouble(),
      isActive: json["isActive"] as bool? ?? true,
      amountDue: (json["amountDue"] as num?)?.toDouble() ?? 0.0,
      creditBalance: (json["creditBalance"] as num?)?.toDouble() ?? 0.0,
    );
  }
}

File: apps/api/src/models/debt.ts
import { AggregateRoot } from "@abejarano/ts-mongodb-criteria"
import { createMongoId } from "./shared/mongo_id"

export enum DebtType {
  CreditCard = "credit_card",
  Loan = "loan",
  Other = "other",
}

export type DebtPrimitives = {
  id?: string
  debtId: string
  userId: string
  name: string
  type: DebtType
  linkedAccountId?: string
  paymentAccountId?: string
  currency: string
  currentBalance: number
  initialBalance?: number
  dueDay?: number
  minimumPayment?: number
  interestRateAnnual?: number
  isActive: boolean
  createdAt: Date
  updatedAt?: Date
  amountDue?: number
  creditBalance?: number
}

export type DebtCreateProps = {
  userId: string
  name: string
  type: DebtType
  linkedAccountId?: string
  paymentAccountId?: string
  currency?: string
  dueDay?: number
  minimumPayment?: number
  interestRateAnnual?: number
  initialBalance?: number
  isActive?: boolean
}

export class Debt extends AggregateRoot {
  private readonly props: DebtPrimitives

  private constructor(props: DebtPrimitives) {
    super()
    this.props = props
  }

  static create(props: DebtCreateProps): Debt {
    const now = new Date()
    return new Debt({
      debtId: createMongoId(),
      userId: props.userId,
      name: props.name,
      type: props.type,
      linkedAccountId: props.linkedAccountId,
      paymentAccountId: props.paymentAccountId,
  currency: props.currency ?? "BRL",
      currentBalance: 0,
      initialBalance: props.initialBalance ?? 0,
      dueDay: props.dueDay,
      minimumPayment: props.minimumPayment,
      interestRateAnnual: props.interestRateAnnual,
      isActive: props.isActive ?? true,
      createdAt: now,
      updatedAt: now,
    })
  }

  static override fromPrimitives(primitives: DebtPrimitives): Debt {
    return new Debt(primitives)
  }

  toPrimitives(): DebtPrimitives {
    return this.props
  }

  getId(): string {
    return this.props.id ?? this.props.debtId
  }

  getDebtId(): string {
    return this.props.debtId
  }
}
